### ISSUE TITLE
[red-knot] Union incorrectly reported as not iterable

### ISSUE THREAD

**Issue by MichaReiser** [astral-sh/ruff#13990]
Given:

```py

class TestIter:
    def __next__(self) -> str:
        return "abcd"


class Test:
    def __iter__(self) -> TestIter:
        return TestIter()


def bool_instance() -> bool:
    return True


flag = bool_instance()

if flag:
    a = "123"
else:
    a = Test()

x = "fallback"
for x in a:
    ...

reveal_type(x)
``` 

**Actual output**

```
ERROR /home/micha/astral/test/src/test.py:26:10: Object of type `Literal["123"] | Test` is not iterable
ERROR /home/micha/astral/test/src/test.py:29:1: Revealed type is `Literal["fallback"] | Unknown`
```

**Expected output**

```
ERROR /home/micha/astral/test/src/test.py:29:1: Revealed type is `Literal["fallback"] | str`
```

As part of this, we should also revisit if passing `self` here is correct because the first argument to `next` in this example isn't `Test` but `TestIter`

https://github.com/astral-sh/ruff/blob/1607d88c2253473a116155176effa4ab14eafcce/crates/red_knot_python_semantic/src/types.rs#L1115-L1121

**Comment by AlexWaygood** [comment]
Here's a slightly more self-contained example (because it doesn't involve any special understanding of the fact that e.g. a string literal is iterable), that makes it very obvious that the issue is the union type, as you say:

```py
from typing_extensions import reveal_type

class TestIter:
    def __next__(self) -> int:
        return 42

class Test:
    def __iter__(self) -> TestIter:
        return TestIter()

class Test2:
    def __iter__(self) -> TestIter:
        return TestIter()

def bool_instance() -> bool:
    return True

flag = bool_instance()
x = y = "fallback"

for x in Test():
    ...

for y in (Test() if flag else Test2()):
    ...

reveal_type(x)
reveal_type(y)
```

red-knot reports:

```
ERROR /Users/alexw/dev/experiment/foo.py:28:11: Object of type `Test | Test2` is not iterable
ERROR /Users/alexw/dev/experiment/foo.py:31:1: Revealed type is `Literal["fallback"] | int`
ERROR /Users/alexw/dev/experiment/foo.py:32:1: Revealed type is `Literal["fallback"] | Unknown`
```
